materialize-mongodb/spec.go

```diff
@@ -0,0 +1,94 @@
+package main
+
+import (
+	"fmt"
+	"context"
+
+	pf "github.com/estuary/flow/go/protocols/flow"
+	proto "github.com/gogo/protobuf/proto"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo"
+)
+
+const specCollection = "flow_materializations"
+
+type SavedSpec struct {
+	Id   string `bson:"uuid"`
+	Spec []byte `bson:"spec"`
+}
+
+// LoadSpec loads existing spec from S3 and create a map of it by table name
+func (d driver) LoadSpec(ctx context.Context, cfg config, materialization string) (map[string]*pf.MaterializationSpec_Binding, error) {
+	var client, err = d.connect(ctx, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("connecting to database: %w", err)
+	}
+
+	var collection = client.Database(cfg.Database).Collection(specCollection)
+
+	var s SavedSpec
+
+	err = collection.FindOne(ctx, bson.D{{idField, bson.D{{"$eq", materialization}}}}).Decode(&s)
+	if err != nil {
+		if err == mongo.ErrNoDocuments {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("could not find spec: %w", err)
+	}
+
+	var existing pf.MaterializationSpec
+	err = proto.Unmarshal(s.Spec, &existing)
+	if err != nil {
+		return nil, fmt.Errorf("parsing existing materialization spec: %w", err)
+	}
+
+	bindingsByTable := make(map[string]*pf.MaterializationSpec_Binding)
+
+	for _, binding := range existing.Bindings {
+		var r resource
```

Comment:
I might be missing a subtlety here, but something feels off about using a map for these. Is it possible to have multiple flow collections materialized into the same mongodb collection? I think not since resource paths have to be unique.

It seems to me that a main challenge is that the `ResourcePath` of the binding from the spec isn't _directly_ usable (I don't think?) in `db.Col` since it includes both the mongodb database and mongodb collection name. But an option would be to index `binding.ResourcePath[1]` to get the mongodb collection name, assuming that the second element in the `ResourcePath` is always the mongodb collection name, which it certainly should be. Then we could return a `[] *pf.MaterializationSpec_Binding` here and in `driver.go` ~line 127 compare the proposed bindings with the existing bindings more directly (without unmarshalling the resource spec) and do the deletion based on `binding.ResourcePath[1]`.